quota/etcd/quotaapi/quota_server_test.go

```diff
@@ -16,78 +16,459 @@ package quotaapi
+		},
+		{
+			desc: "unknownConfig",
+			req: &quotapb.UpdateConfigRequest{
+				Name:       globalWriteName,
+				Config:     &quotapb.Config{MaxTokens: 200},
+				UpdateMask: &field_mask.FieldMask{Paths: []string{"max_tokens"}},
+			},
+			wantCode: codes.NotFound,
+		},
+		{
+			desc:    "badConfig",
+			baseCfg: globalWrite,
+			req: &quotapb.UpdateConfigRequest{
+				Name:       globalWriteName,
+				Config:     &quotapb.Config{}, // State == UNKNOWN
+				UpdateMask: &field_mask.FieldMask{Paths: []string{"state"}},
+			},
+			// TODO(codingllama): Surface the appropriate error codes from storage
+			wantCode: codes.Unknown,
+			wantCfg:  globalWrite,
+		},
+		{
+			desc:    "badPath",
+			baseCfg: globalWrite,
+			req: &quotapb.UpdateConfigRequest{
+				Name:       globalWriteName,
+				Config:     &quotapb.Config{},
+				UpdateMask: &field_mask.FieldMask{Paths: []string{"NOT_A_FIELD"}},
+			},
+			wantCode: codes.InvalidArgument,
+			wantCfg:  globalWrite,
+		},
+		{
+			desc: "emptyName",
+			req: &quotapb.UpdateConfigRequest{
+				Config:     &quotapb.Config{MaxTokens: 100},
+				UpdateMask: &field_mask.FieldMask{Paths: []string{"max_tokens"}},
+			},
+			wantCode: codes.InvalidArgument,
+		},
+		{
+			desc:    "emptyConfig",
+			baseCfg: globalWrite,
+			req: &quotapb.UpdateConfigRequest{
+				Name:       globalWriteName,
+				UpdateMask: &field_mask.FieldMask{Paths: []string{"max_tokens"}},
+			},
+			wantCode: codes.InvalidArgument,
+			wantCfg:  globalWrite,
+		},
+		{
+			desc:    "emptyMask",
+			baseCfg: globalWrite,
+			req: &quotapb.UpdateConfigRequest{
+				Name:   globalWriteName,
+				Config: &quotapb.Config{MaxTokens: 100},
+			},
+			wantCode: codes.InvalidArgument,
+			wantCfg:  globalWrite,
+		},
+		{
+			desc:    "emptyConfigWithReset",
+			baseCfg: globalWrite,
+			req: &quotapb.UpdateConfigRequest{
+				Name:       globalWriteName,
+				UpdateMask: &field_mask.FieldMask{Paths: []string{"max_tokens"}},
+				ResetQuota: true,
+			},
+			wantCode: codes.InvalidArgument,
+			wantCfg:  globalWrite,
+		},
+		{
+			desc:    "emptyMaskWithReset",
+			baseCfg: globalWrite,
+			req: &quotapb.UpdateConfigRequest{
+				Name:       globalWriteName,
+				Config:     &quotapb.Config{MaxTokens: 100},
+				ResetQuota: true,
+			},
+			wantCode: codes.InvalidArgument,
+			wantCfg:  globalWrite,
 		},
 	}
 
 	ctx := context.Background()
-	want := codes.Unimplemented
+	rpc := func(ctx context.Context, req interface{}) (*quotapb.Config, error) {
+		return quotaClient.UpdateConfig(ctx, req.(*quotapb.UpdateConfigRequest))
+	}
 	for _, test := range tests {
-		err := test.fn(ctx, client)
-		switch s, ok := status.FromError(err); {
-		case !ok:
-			t.Errorf("%v() returned a non-gRPC error: %v", test.desc, err)
-		case s.Code() != want:
-			t.Errorf("%v() returned err = %q, want code = %s", test.desc, err, want)
+		if err := runUpsertTest(ctx, test, rpc, "UpdateConfig"); err != nil {
+			t.Error(err)
+		}
+	}
+}
+
```

Comment:
Would it be possible to add a test that updates a large number of configs in parallel on different goroutines, some with the same name and some different, so that if there are any thread-safety issues we've got a chance of catching them with the `-race` runs